File: scripts/dual_scraper.py
import requests
from bs4 import BeautifulSoup
import trafilatura
import os
import json
import time

LINKS_FILE = "data/master_links_live.txt"
BS4_DIR = "data/dual_scrape/bs4"
TRAF_DIR = "data/dual_scrape/trafilatura"
UNIFIED_DIR = "data/dual_scrape/unified"

def clean_filename(url):
    return url.split("/")[-1].replace(".php", "").replace(".html", "") or "index"

def scrape_bs4(url):
    try:
        response = requests.get(url, timeout=15)
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Remove script and style elements
        for script in soup(["script", "style"]):
            script.extract()

        # Get text with better formatting
        lines = (line.strip() for line in soup.get_text().splitlines())
        chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
        text = '\n'.join(chunk for chunk in chunks if chunk)
        
        # Specifically get tables
        tables = soup.find_all('table')
        table_data = ""
        for i, table in enumerate(tables):
            table_data += f"\n--- TABLE {i+1} ---\n"
            for row in table.find_all('tr'):
                cells = [cell.get_text().strip() for cell in row.find_all(['td', 'th'])]
                table_data += " | ".join(cells) + "\n"
        
        return f"URL: {url}\n\nSTRUCTURED TEXT:\n{text}\n\nTABLE DATA:\n{table_data}"
    except Exception as e:
        return f"BS4 Error: {e}"

def scrape_trafilatura(url):
    try:
        downloaded = trafilatura.fetch_url(url)
        content = trafilatura.extract(downloaded, include_tables=True, include_links=True, output_format='markdown')
        return f"URL: {url}\n\nMARKDOWN CONTENT:\n{content}"
    except Exception as e:
        return f"Trafilatura Error: {e}"

def main():
    if not os.path.exists(LINKS_FILE):
        print(f"Error: {LINKS_FILE} not found.")
        return

    with open(LINKS_FILE, "r") as f:
        links = [l.strip() for l in f.readlines() if l.strip()]

    print(f"Starting Dual Scrape for {len(links)} links...")

    for url in links:
        fname = clean_filename(url)
        print(f"Scraping: {url} -> {fname}")
        
        # Method 1
        bs4_data = scrape_bs4(url)
        with open(f"{BS4_DIR}/{fname}.txt", "w", encoding="utf-8") as f:
            f.write(bs4_data)
            
        # Method 2
        traf_data = scrape_trafilatura(url)
        with open(f"{TRAF_DIR}/{fname}.md", "w", encoding="utf-8") as f:
            f.write(traf_data if traf_data else "No content extracted.")
            
        # Simple Merge (Initial version - deduplicating by combining and letting the RAG handle context)
        unified_content = f"--- SOURCE: {url} ---\n\n"
        unified_content += "[[ STRUCTURED DATA (BS4) ]]\n"
        unified_content += bs4_data + "\n\n"
        unified_content += "[[ NARRATIVE DATA (TRAFILATURA) ]]\n"
        unified_content += (traf_data if traf_data else "[No narrative content]")
        
        with open(f"{UNIFIED_DIR}/{fname}.unified.txt", "w", encoding="utf-8") as f:
            f.write(unified_content)
        
        time.sleep(1) # Polite scraping

    print("Dual Scrape Complete.")

if __name__ == "__main__":
    main()

File: scripts/dual-scraper.ts
import axios from 'axios';
import * as cheerio from 'cheerio';
import fs from 'fs';
import path from 'path';

const LINKS_FILE = 'data/architecture/master_links.txt';
const DIR_CRAWLER = 'data/raw_crawler';
const DIR_DETAILED = 'data/raw_detailed';

async function dualScrape() {
    const links = fs.readFileSync(LINKS_FILE, 'utf-8')
        .split('\n')
        .map(l => l.trim())
        .filter(l => l && l.startsWith('http'));

    console.log(`🚀 Starting dual-method scrape for ${links.length} URLs...`);

    for (let i = 0; i < links.length; i++) {
        const url = links[i];
        const filename = url.split('/').pop()?.replace('.php', '') || 'index';
        const finalName = filename === 'index' || filename === '' ? 'home' : filename;

        console.log(`[${i+1}/${links.length}] Scraping: ${url}...`);

        try {
            const { data: html } = await axios.get(url, {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
                },
                timeout: 15000
            });

            const $ = cheerio.load(html);

            // Removing ONLY scripts and styles
            $('script, style').remove();

            // Method 1: Semantic Markdown-like Crawler
            let crawlerText = `URL: ${url}\nTITLE: ${$('title').text()}\n\n`;
            $('h1, h2, h3, h4, h5, p, li').each((_, el) => {
                const text = $(el).text().trim().replace(/\s+/g, ' ');
                if (text.length > 5) {
                    crawlerText += `${text}\n`;
                }
            });

            // Method 2: Table-Rich Detailed Extractor
            let detailedText = `SOURCE: ${url}\n\n`;
            
            // Extract Tables
            $('table').each((tIdx, table) => {
                detailedText += `\n[TABLE_${tIdx+1}]\n`;
                $(table).find('tr').each((_, tr) => {
                    const row: string[] = [];
                    $(tr).find('th, td').each((_, td) => {
                        row.push($(td).text().trim().replace(/\s+/g, ' '));
                    });
                    if (row.length > 0) detailedText += `| ${row.join(' | ')} |\n`;
                });
            });

            // Extract Body Text (Brute force but safe)
            const bodyContent = $('body').text().trim().replace(/\s+/g, ' ');
            detailedText += `\n\n[FULL_BODY_TEXT]\n${bodyContent}`;

            // Saving
            fs.writeFileSync(path.join(DIR_CRAWLER, `${finalName}.crawler.txt`), crawlerText.trim());
            fs.writeFileSync(path.join(DIR_DETAILED, `${finalName}.detailed.txt`), detailedText.trim());

            console.log(`✅ Saved: ${finalName} (${detailedText.length} bytes)`);



        } catch (err: any) {
            console.error(`❌ Failed: ${url} -> ${err.message}`);
        }

        // Polite delay
        await new Promise(r => setTimeout(r, 200));
    }

    console.log('✨ Dual-method scraping completed!');
}

dualScrape().catch(console.error);
